File: ccunningham_04.py
from string import digits


if __name__ == '__main__':
    possible = (str(v) for v in range(130254, 678275))

    possible_count = 0
    filtered_count = 0

    for v in possible:
        if ''.join(sorted(v)) != v or not any(v.count(d) > 1 for d in digits):
            continue

        possible_count += 1
        if any(v.count(d) == 2 for d in digits):
            filtered_count += 1

    print(f'Part 01: {possible_count}')
    print(f'Part 02: {filtered_count}')

File: ccunningham_04.go
package main

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
)

var digits = []rune("0123456789")

func main() {
	possibleCount := 0
	filteredCount := 0

	for i := 130254; i < 678275; i++ {
		v := strconv.Itoa(i)

		vSorted := []rune(v); sort.Slice(vSorted, func(i, j int) bool {
			return vSorted[i] < vSorted[j]
		})

		if v != string(vSorted) {
			continue
		}

		hasPair := false
		for _, d := range digits {
			if strings.Count(v, string(d)) > 1 {
				hasPair = true
				break
			}
		}

		if hasPair {
			possibleCount++
		}

		for _, d := range digits {
			if strings.Count(v, string(d)) == 2 {
				filteredCount++
				break
			}
		}
	}

	fmt.Printf("Part 01: %v\n", possibleCount)
	fmt.Printf("Part 02: %v\n", filteredCount)
}
